### PR TITLE
Omnibus to implement CFRunLoop with kevent.

### PR DESCRIPTION
**Pull request by 3405691582**
Dispatch uses kevent for its event loop backend. However, `CFRunLoop` in Foundation uses Mach ports. This means for other kevent-based platforms such as OpenBSD, new runloop functionality is required, which this omnibus pr implements. 

This pr must be considered in concert with apple/swift-corelibs-libdispatch#559 (see below).

Please feel free to request any or all to be split into multiple prs, though all these commits and the aforementioned Dispatch pr are required to properly implement `CFRunLoop` on this platform.

Notes:
* Some of the Mach-based implementation refers to `MACH_PORT_NULL`, for example, when we have symbols already defined like `CFPORT_NULL`. These are brought more into alignment for consistency.
* Each platform necessitates implementing the "backend" set of `__CFPort*` and `mk_timer_*` stubs. It does not make sense if these methods are not implemented for a given platform, so add a default `#error` case. (The code would greatly benefit from reorganizing the necessary stubs to implement together in one `#if` conditional, but this should probably be done separate from this omnibus as it is not critically important to achieve our result. However, we have implemented all the stubs together below in this commit.)
* A tentative kevent-based implementation of these stubs are provided. In apple/swift-corelibs-libdispatch#559, we introduced a runloop implementation using `pipe2` and packing each 32-bit fd in the pipe into a 64-bt integer. Since CFRunLoop fraternizes with Dispatch via `_dispatch_main_queue_callback_4CF`, this means that `dispatch_runloop_handle_t` in Dispatch must match `__CFPort` here. Unfortunately, this means that we have to perform some bit-packing shenanigans to handle kevent timers. There might be nicer ways to do this, but we'll defer that for now.
* We need to fix SR-14288 while we are here to ensure the runloop timers work correctly and the tests pass properly.
* While implementing the runloop, testing has uncovered some strange behavior with `pthread_main_np` and TSD on this platform. Work around this by doing a simple pthread_self comparison.

### CODE REVIEW

**Comment by 3405691582** [comment]
cc @millenomi -- input or advice on how to proceed?

**Comment by AtariDreams** [comment]
@compnerd any thoughts?

**Comment by buttaface** [comment]
@parkera, mind reviewing?

**Comment by buttaface** [comment]
@phausler, maybe you could review?

**Comment by buttaface** [comment]
@etcwilde, would you review?

File: CoreFoundation/RunLoop.subproj/CFRunLoop.c
Comment (etcwilde) [inline CoreFoundation/RunLoop.subproj/CFRunLoop.c:54]: @compnerd, is this the appropriate implementation of this function for Windows?

File: CoreFoundation/Base.subproj/CFPlatform.c
Comment (etcwilde) [inline CoreFoundation/Base.subproj/CFPlatform.c:252]: It's a bit clearer if we have the OpenBSD case come first and not invert the condition.
```
#if defined(__OpenBSD__) // special case for OpenBSD
    return pthread_equal(pthread_self(), _CFMainPThread) != 0;
#else // everything else
    return pthread_main_np() == 1;
#endif
```
I usually think of `else` indicating anything not special-cased, so having the else only catch the OpenBSD case seems a bit odd.
Thoughts?

File: CoreFoundation/RunLoop.subproj/CFRunLoop.c
Comment (etcwilde) [inline CoreFoundation/RunLoop.subproj/CFRunLoop.c:574]: Inconsistent indentation through here.

**Review by etcwilde — COMMENTED** [review]
This looks fine for the most part. I left a couple comments on style.

I'm unfortunately not familiar enough with this part of the codebase to really identify logic/behavior issues in the implementation of the `__CFPort*` or `mk_timer` function implementations down in `CFRunLoop.c`. It looks fine from what I can tell, but it would be better to get someone more familiar with this code to look over it and verify that it's correct.

As I'm not very familiar with this part of the codebase, nor am I an owner or maintainer of foundation, I don't really feel comfortable just approving it. I'll leave that part up to @parkera.

File: CoreFoundation/Base.subproj/CFPlatform.c
Comment (3405691582) [reply CoreFoundation/Base.subproj/CFPlatform.c]: Reordered the conditional.

File: CoreFoundation/RunLoop.subproj/CFRunLoop.c
Comment (3405691582) [reply CoreFoundation/RunLoop.subproj/CFRunLoop.c]: Fixed.

**Review by parkera — APPROVED** [review]
Looks good to me.

**Comment by parkera** [comment]
@swift-ci test

**Comment by compnerd** [comment]
I think that this might be causing the timeout on Windows ...

**Pull request merged**
